### ISSUE TITLE
Handle entity/kind schema changes/versioning

### ISSUE THREAD

**Issue by joshua-berry-ntnx** [nutanix/papiea-js#349]
We need a way to version spec/status data handled by Papiea as we make schema changes over time.

Could we leverage the same intentful engine for driving schema upgrades? For example, set the schema version in the spec to N+1, and then use the intentful engine to upgrade the schema if we see a status with an older schema version.

**Comment by Emulebest** [comment]
Can you provide a more particular example of this issue?

**Comment by joshua-berry-ntnx** [comment]
I didn't have a particular example in mind; this was more of a, "how can we make sure we can change the definitions of entities after the fact when we are running in production?" Some things we might want to do:

- Adding fields
- Removing fields
- Renaming fields
- Changing the type of a field
- Applying some migration/script to change the semantic meaning of a particular field/fields

We don't necessarily have to support ALL of those things, but we should probably at least have a plan for what we will support and how it will work.

**Comment by shlomi-nx** [comment]
Yeah, this is kinda like django's db versions. However, this may actually be a bit more involved, since this update may be hot - it may happen as the service and papiea are running.

We need to think how can one upgrade a version while the previous version is in use. Binny and I had a few brainstorming sessions on this, lets discuss it.

**Comment by joshua-berry-ntnx** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Address Papiea issues needed for inventory-only BMaaS](https://trello.com/c/V6gmTU73/66-address-papiea-issues-needed-for-inventory-only-bmaas)

**Comment by joshua-berry-ntnx** [comment]
Edited the description to add some thoughts on minimum requirements for a 1.0 release.
